File: src/theme/derive.ts
import type { ColorScheme, UiVars } from './tokens'

/** Ensure #rrggbb (CSS vars need #). */
export function normalizeHex(color: string | undefined, fallback: string): string {
  if (!color) return fallback
  const c = color.trim()
  if (c.startsWith('#')) return c.length === 4 ? expandShortHex(c) : c.slice(0, 7)
  if (/^[0-9a-fA-F]{6}$/.test(c)) return `#${c}`
  if (/^[0-9a-fA-F]{3}$/.test(c)) return expandShortHex(`#${c}`)
  return fallback
}

function expandShortHex(short: string): string {
  const r = short[1]
  const g = short[2]
  const b = short[3]
  return `#${r}${r}${g}${g}${b}${b}`
}

function parseRgb(hex: string): [number, number, number] {
  const h = normalizeHex(hex, '#000000').slice(1)
  return [parseInt(h.slice(0, 2), 16), parseInt(h.slice(2, 4), 16), parseInt(h.slice(4, 6), 16)]
}

function toHex(r: number, g: number, b: number): string {
  const clamp = (n: number) => Math.max(0, Math.min(255, Math.round(n)))
  return `#${[clamp(r), clamp(g), clamp(b)].map((n) => n.toString(16).padStart(2, '0')).join('')}`
}

/** Mix `a` toward `b` by t in [0,1]. */
export function mix(a: string, b: string, t: number): string {
  const [ar, ag, ab] = parseRgb(a)
  const [br, bg, bb] = parseRgb(b)
  return toHex(ar + (br - ar) * t, ag + (bg - ag) * t, ab + (bb - ab) * t)
}

export function lighten(hex: string, t: number): string {
  return mix(hex, '#ffffff', t)
}

export function darken(hex: string, t: number): string {
  return mix(hex, '#000000', t)
}

type ColorMap = Record<string, string>

/**
 * Map VS Code–style workbench/editor colors into Conflicto UI CSS variables.
 * Missing keys are derived from background/foreground so every theme stays complete.
 */
export function deriveUiVars(colors: ColorMap, scheme: ColorScheme): UiVars {
  const isLight = scheme === 'light'
  const bg = normalizeHex(colors['editor.background'], isLight ? '#ffffff' : '#1e1e1e')
  const fg = normalizeHex(colors['editor.foreground'], isLight ? '#000000' : '#cccccc')
  const sidebar = normalizeHex(
    colors['sideBar.background'] ?? colors['editorWidget.background'],
    isLight ? darken(bg, 0.03) : lighten(bg, 0.04),
  )
  const surfaceFallback = isLight ? darken(bg, 0.04) : lighten(bg, 0.06)
  let surface = normalizeHex(
    colors['editorWidget.background'] ?? colors['panel.background'],
    surfaceFallback,
  )
  // Keep chrome strips readable when widget bg matches sidebar or editor.
  if (surface.toLowerCase() === bg.toLowerCase() || surface.toLowerCase() === sidebar.toLowerCase()) {
    surface = isLight ? darken(sidebar, 0.04) : lighten(sidebar, 0.05)
    if (surface.toLowerCase() === sidebar.toLowerCase() || surface.toLowerCase() === bg.toLowerCase()) {
      surface = isLight ? darken(bg, 0.06) : lighten(bg, 0.08)
    }
  }
  const border = normalizeHex(
    colors['editorWidget.border'] ?? colors['panel.border'],
    isLight ? darken(bg, 0.12) : lighten(bg, 0.12),
  )
  const muted = normalizeHex(colors['editorLineNumber.foreground'], mix(fg, bg, 0.45))
  const accent = normalizeHex(colors['focusBorder'] ?? colors['button.background'], '#0078d4')
  const accentHover = lighten(accent, isLight ? 0.08 : 0.12)
  const btnBg = normalizeHex(
    colors['button.secondaryBackground'] ?? colors['input.background'],
    isLight ? darken(bg, 0.06) : lighten(bg, 0.1),
  )
  const btnHover = isLight ? darken(btnBg, 0.08) : lighten(btnBg, 0.08)
  const hover = normalizeHex(colors['list.hoverBackground'], mix(bg, fg, isLight ? 0.06 : 0.08))
  const active = normalizeHex(colors['list.activeSelectionBackground'], mix(bg, fg, isLight ? 0.1 : 0.14))

  const statusM = normalizeHex(colors['gitDecoration.modifiedResourceForeground'], '#e2c08d')
  const statusA = normalizeHex(colors['gitDecoration.addedResourceForeground'], '#73c991')
  const statusD = normalizeHex(colors['gitDecoration.deletedResourceForeground'], '#f14c4c')
  const statusR = normalizeHex(colors['gitDecoration.renamedResourceForeground'], statusA)

  const dangerBorder = normalizeHex(colors['errorForeground'] ?? colors['inputValidation.errorBorder'], '#f14c4c')
  const dangerBg = mix(bg, dangerBorder, isLight ? 0.12 : 0.28)
  const dangerFg = isLight ? darken(dangerBorder, 0.25) : lighten(dangerBorder, 0.35)

  const refFg = normalizeHex(colors['textLink.foreground'], accentHover)
  const refBg = mix(bg, accent, isLight ? 0.12 : 0.22)

  return {
    '--bg': bg,
    '--bg-sidebar': sidebar,
    '--bg-surface': surface,
    '--bg-hover': hover,
    '--bg-active': active,
    '--border': border,
    '--text': fg,
    '--text-muted': muted,
    '--accent': accent,
    '--accent-hover': accentHover,
    '--btn-bg': btnBg,
    '--btn-hover': btnHover,
    '--btn-fg': '#ffffff',
    '--status-m': statusM,
    '--status-a': statusA,
    '--status-d': statusD,
    '--status-r': statusR,
    '--danger-bg': dangerBg,
    '--danger-border': dangerBorder,
    '--danger-fg': dangerFg,
    '--ref-fg': refFg,
    '--ref-bg': refBg,
  }
}

File: native/conflicto_core/src/theme/derive.rs
use super::color::{darken, hex_to_rgb, lighten, mix};
use super::{ColorScheme, UiVars};

fn get<'a>(colors: &'a [(&str, &str)], key: &str) -> Option<&'a str> {
    colors.iter().find(|(k, _)| *k == key).map(|(_, v)| *v)
}

pub fn derive_ui_vars(colors: &[(&str, &str)], scheme: ColorScheme) -> UiVars {
    let is_light = matches!(scheme, ColorScheme::Light);
    let bg = hex_to_rgb(
        get(colors, "editor.background"),
        if is_light { "#ffffff" } else { "#1e1e1e" },
    );
    let fg = hex_to_rgb(
        get(colors, "editor.foreground"),
        if is_light { "#000000" } else { "#cccccc" },
    );
    let sidebar = match get(colors, "sideBar.background").or_else(|| get(colors, "editorWidget.background"))
    {
        Some(c) => hex_to_rgb(Some(c), "#000000"),
        None => {
            if is_light {
                darken(bg, 0.03)
            } else {
                lighten(bg, 0.04)
            }
        }
    };
    let surface_fallback = if is_light {
        darken(bg, 0.04)
    } else {
        lighten(bg, 0.06)
    };
    let mut surface = match get(colors, "editorWidget.background").or_else(|| get(colors, "panel.background"))
    {
        Some(c) => hex_to_rgb(Some(c), "#000000"),
        None => surface_fallback,
    };
    if surface == bg || surface == sidebar {
        surface = if is_light {
            darken(sidebar, 0.04)
        } else {
            lighten(sidebar, 0.05)
        };
        if surface == sidebar || surface == bg {
            surface = if is_light {
                darken(bg, 0.06)
            } else {
                lighten(bg, 0.08)
            };
        }
    }
    let border = match get(colors, "editorWidget.border").or_else(|| get(colors, "panel.border")) {
        Some(c) => hex_to_rgb(Some(c), "#000000"),
        None => {
            if is_light {
                darken(bg, 0.12)
            } else {
                lighten(bg, 0.12)
            }
        }
    };
    let muted = match get(colors, "editorLineNumber.foreground") {
        Some(c) => hex_to_rgb(Some(c), "#000000"),
        None => mix(fg, bg, 0.45),
    };
    let accent = hex_to_rgb(
        get(colors, "focusBorder").or_else(|| get(colors, "button.background")),
        "#0078d4",
    );
    let accent_hover = lighten(accent, if is_light { 0.08 } else { 0.12 });
    let btn_bg = match get(colors, "button.secondaryBackground").or_else(|| get(colors, "input.background"))
    {
        Some(c) => hex_to_rgb(Some(c), "#000000"),
        None => {
            if is_light {
                darken(bg, 0.06)
            } else {
                lighten(bg, 0.1)
            }
        }
    };
    let btn_hover = if is_light {
        darken(btn_bg, 0.08)
    } else {
        lighten(btn_bg, 0.08)
    };
    let hover = match get(colors, "list.hoverBackground") {
        Some(c) => hex_to_rgb(Some(c), "#000000"),
        None => mix(bg, fg, if is_light { 0.06 } else { 0.08 }),
    };
    let active = match get(colors, "list.activeSelectionBackground") {
        Some(c) => hex_to_rgb(Some(c), "#000000"),
        None => mix(bg, fg, if is_light { 0.1 } else { 0.14 }),
    };
    let status_m = hex_to_rgb(
        get(colors, "gitDecoration.modifiedResourceForeground"),
        "#e2c08d",
    );
    let status_a = hex_to_rgb(
        get(colors, "gitDecoration.addedResourceForeground"),
        "#73c991",
    );
    let status_d = hex_to_rgb(
        get(colors, "gitDecoration.deletedResourceForeground"),
        "#f14c4c",
    );
    let status_r = match get(colors, "gitDecoration.renamedResourceForeground") {
        Some(c) => hex_to_rgb(Some(c), "#000000"),
        None => status_a,
    };
    let danger_border = hex_to_rgb(
        get(colors, "errorForeground").or_else(|| get(colors, "inputValidation.errorBorder")),
        "#f14c4c",
    );
    let danger_bg = mix(bg, danger_border, if is_light { 0.12 } else { 0.28 });
    let danger_fg = if is_light {
        darken(danger_border, 0.25)
    } else {
        lighten(danger_border, 0.35)
    };
    let ref_fg = match get(colors, "textLink.foreground") {
        Some(c) => hex_to_rgb(Some(c), "#000000"),
        None => accent_hover,
    };
    let ref_bg = mix(bg, accent, if is_light { 0.12 } else { 0.22 });

    UiVars {
        bg,
        bg_sidebar: sidebar,
        bg_surface: surface,
        bg_hover: hover,
        bg_active: active,
        border,
        text: fg,
        text_muted: muted,
        accent,
        accent_hover,
        btn_bg,
        btn_hover,
        btn_fg: [255, 255, 255],
        status_m,
        status_a,
        status_d,
        status_r,
        danger_bg,
        danger_border,
        danger_fg,
        ref_fg,
        ref_bg,
    }
}
